awscli/errorhandler.py

```diff
@@ -11,68 +11,110 @@
-    with an appropriate error code.
-    """
-
-    def __call__(self, http_response, parsed, model, **kwargs):
-        LOG.debug('HTTP Response Code: %d', http_response.status_code)
-        error_type = None
-        error_class = None
-        if http_response.status_code >= 500:
-            error_type = 'server'
-            error_class = ServerError
-        elif http_response.status_code >= 400 or http_response.status_code == 301:
-            error_type = 'client'
-            error_class = ClientError
-        if error_class is not None:
-            code, message = self._get_error_code_and_message(parsed)
-            raise error_class(
-                error_code=code, error_message=message,
-                error_type=error_type, operation_name=model.name,
-                http_status_code=http_response.status_code)
-
-    def _get_error_code_and_message(self, response):
-        code = 'Unknown'
-        message = 'Unknown'
-        if 'Error' in response:
-            error = response['Error']
-            return error.get('Code', code), error.get('Message', message)
-        return (code, message)
+class BaseExceptionHandler:
+    def handle_exception(self, exception, stdout, stderr):
+        raise NotImplementedError('handle_exception')
+
+
+class FilteredExceptionHandler(BaseExceptionHandler):
+    EXCEPTIONS_TO_HANDLE = ()
+    MESSAGE = '%s'
+
+    def handle_exception(self, exception, stdout, stderr):
+        if issubclass(exception.__class__, self.EXCEPTIONS_TO_HANDLE):
+            return_val = self._do_handle_exception(exception, stdout, stderr)
+            if return_val is not None:
+                return return_val
+
+    def _do_handle_exception(self, exception, stdout, stderr):
+        stderr.write("\n")
+        stderr.write(self.MESSAGE % exception)
+        stderr.write("\n")
+        return self.RC
+
+
+class ParamValidationErrorsHandler(FilteredExceptionHandler):
+    EXCEPTIONS_TO_HANDLE = (
+        ParamError, ParamSyntaxError, ArgParseException,
+        ParamValidationError, BotocoreParamValidationError,
+        UnknownArgumentError
+    )
+    RC = PARAM_VALIDATION_ERROR_RC
+
+
+class ClientErrorHandler(FilteredExceptionHandler):
+    EXCEPTIONS_TO_HANDLE = ClientError
+    RC = CLIENT_ERROR_RC
+
+
+class ConfigurationErrorHandler(FilteredExceptionHandler):
+    EXCEPTIONS_TO_HANDLE = ConfigurationError
+    RC = CONFIGURATION_ERROR_RC
+
+
+class NoRegionErrorHandler(FilteredExceptionHandler):
+    EXCEPTIONS_TO_HANDLE = NoRegionError
+    RC = CONFIGURATION_ERROR_RC
+    MESSAGE = '%s You can also configure your region by running "aws configure".'
+
+
+class NoCredentialsErrorHandler(FilteredExceptionHandler):
+    EXCEPTIONS_TO_HANDLE = NoCredentialsError
+    RC = CONFIGURATION_ERROR_RC
+    MESSAGE = '%s. You can configure credentials by running "aws configure".'
+
+
+class UnknownArgumentErrorHandler(FilteredExceptionHandler):
+    EXCEPTIONS_TO_HANDLE = UnknownArgumentError
```

Comment:
Is there a reason the new `ArgParseException` is not handled here as well? It seems like we can consolidate this class into `ParamValidationErrorsHandler` or delegate some of the error formatting logic in the updates for the `ArgParseException` class to this handler.

Reply:
Parsers use `format_usage` method to create usage message for `ArgParseException` and  different parsers can use different formats or pass there more than just a `USAGE`